### DISCUSSION TITLE
Filter buttons stacking filters instead of switching them

### DISCUSSION THREAD

**Discussion by robertncole1** [nss-evening-cohort-14/evening-client-side#280] [Help Me!] [answered]
Name: Rob Cole
I have dynamically created filter buttons that will filter my API search results. On the first click, the filter works! When choosing a new filter, it filters on top of that first filter, instead of filtering on the original search results. I tried to create a clear button but it still isn't working.

List any error messages that appear:
I am getting some errors on proptypes but that is from the API results returning some arrays and some strings. They come up regardless of the filter buttons. So no relevant errors.

What have you googled?:
I have been using a tutorial for these filter buttons, but mine doesn't seem to work like his: https://www.youtube.com/watch?v=cbEHLalLLeM
https://stackoverflow.com/questions/57358605/multiple-filters-in-react
https://stackoverflow.com/questions/60279994/how-to-clear-all-filters-in-react-table

What have you tried? Tell us everything you have tried:
I have tried setting up my buttons in a number of different ways. I thought maybe it was because I was dynamically pulling in the buttons based on what the API was returning but when I changed them to static buttons based on type, it did the same thing. The clear button is a solution so I tried setting the setResults(results) as the clear button, thinking that would just return the results, but it just returns a blank page.

Any API Responses you are getting:
I am getting an API response on the initial search, then getting a filtered response when clicking for the first time, and then noting because the filters are stacking.

Copy/paste link TO THE FILE on your branch or in your repo:
Where the results are being filtered: https://github.com/robertncole1/sound-selection/blob/main/src/Components/FilterButtons.js

**Comment by TrinityChristiana** [comment]
Each time you click the button, it looks at the already filtered results and filters those again [here](https://github.com/robertncole1/sound-selection/blob/0af90e992bcf9dc09087c96c93b34027775e13d1/src/Components/FilterButtons.js#L19). 

Try storing two categories of results in two different variables/states. One that stores all results unfiltered (will only get updated when the API call is made) and one that stores the filtered results (this is the one you will update on button click). 

So whenever you filter, you always have access to the unfiltered results, and still have somewhere to store the filtered results

**Reply by robertncole1** [reply]
Maybe I am not following, but when I set two variables/states to an unfiltered view, it isn't carrying over the original search parameter that I put in the input. Its running the api call without the "apiInput" I am plugging into the call. I tried passing that down as a prop that didnt work.

**Comment by robertncole1** [comment, marked as answer]
So my issues ticket changed a bit because I am dynamically pulling in the buttons based on the type return from my api. Instead of having filters statically there, I had a "return to search results button." This was not working because my results state was being set in my app index and then I was doing some crazy prop drilling to make it all happen. This was causing my app to remount and not keeping the state set in the search. Once we moved results to the proper file, it solved some issues. On top of that, we had to create a useEffect in my FIlterButtons so that we were defining the setButtons state and passing a dependency array to watch for results in that useEffect. Ticket number #238 was helpful!
